README
About MoritzServer
==================

When I bought some thermostats to implement a first step in smart home, there were only the options to either use a proprietary remote-controlled LAN-Gateway called MAX!Cube or use FHEM which is both complex and not simple to setup on a NAS.

So a smaller version around a CUL like FHEM supports was implemented from scratch. Idea is to just do data-keeping and leaving the interface part to some other place. Think of it as a gateway device you control yourself.

Prerequisites are a CUL USB stick and a recent CULFW (1.61 or newer) on it.


CHANGELOG
MoritzServer Changelog
======================

Here you can see the full list of changes between each Flask release.

Version 1.0
-----------

Released on Dec 7th 2014

First public release.
Contains simple pairing and set command support


moritzprotocol/test_messages.py
from datetime import datetime
import unittest
from .messages import *


class MessageSampleInputTestCase(unittest.TestCase):
	def test_thermostat_state(self):
		sample = "Z0F61046008FFE90000000019002000CA"
		msg = MoritzMessage.decode_message(sample)
		self.assertTrue(isinstance(msg, ThermostatStateMessage))
		self.assertEqual(msg.counter, 0x61)
		self.assertEqual(msg.flag, 0x4)
		self.assertEqual(msg.sender_id, 0x8FFE9)
		self.assertEqual(msg.receiver_id, 0x0)
		self.assertEqual(msg.group_id, 0)
		self.assertEqual(msg.payload, '19002000CA')
		self.assertEqual(msg.decoded_payload, {
			'battery_low': False,
			'desired_temperature': 16.0,
			'dstsetting': False,
			'is_locked': False,
			'langateway': True,
			'measured_temperature': 20.2,
			'mode': 'manual',
			'rferror': False,
			'valve_position': 0
		})

	def test_set_temperature(self):
		sample = "Z0BB900401234560B3554004B"
		msg = MoritzMessage.decode_message(sample)
		self.assertTrue(isinstance(msg, SetTemperatureMessage))
		self.assertEqual(msg.counter, 0xB9)
		self.assertEqual(msg.flag, 0x0)
		self.assertEqual(msg.sender_id, 0x123456)
		self.assertEqual(msg.receiver_id, 0x0B3554)
		self.assertEqual(msg.group_id, 0)
		self.assertEqual(msg.payload, '4B')
		self.assertEqual(msg.decoded_payload, {
			'desired_temperature': 5.5,
			'mode': 'manual',
		})

	def test_set_temp_ack(self):
		sample = "Z0EB902020B3554123456000119000B"
		msg = MoritzMessage.decode_message(sample)
		self.assertTrue(isinstance(msg, AckMessage))
		self.assertEqual(msg.counter, 0xB9)
		self.assertEqual(msg.flag, 0x02)
		self.assertEqual(msg.sender_id, 0x0B3554)
		self.assertEqual(msg.receiver_id, 0x123456)
		self.assertEqual(msg.group_id, 0)
		self.assertEqual(msg.payload, '0119000B')
		self.assertEqual(msg.decoded_payload, {
			'battery_low': False,
			'desired_temperature': 5.5,
			'dstsetting': False,
			'is_locked': False,
			'langateway': True,
			'mode': 'manual',
			'rferror': False,
			'state': 'ok',
			'valve_position': 0,
		})

	def test_pair_ping(self):
		sample = "Z170004000E016C000000001001A04B455130393932343736"
		msg = MoritzMessage.decode_message(sample)
		self.assertTrue(isinstance(msg, PairPingMessage))
		self.assertEqual(msg.counter, 0x0)
		self.assertEqual(msg.flag, 0x04)
		self.assertEqual(msg.sender_id, 0xE016C)
		self.assertEqual(msg.receiver_id, 0x0)
		self.assertEqual(msg.group_id, 0)
		self.assertEqual(msg.decoded_payload, {
			'firmware_version': "V1.0",
			'device_type': "HeatingThermostat",
			'selftest_result': 0xA0,
			'pairmode': "pair",
			'device_serial': "KEQ0992476",
		})

	def test_pair_pong(self):
		sample = "Z0B0100011234560E016C0000"
		msg = MoritzMessage.decode_message(sample)
		self.assertTrue(isinstance(msg, PairPongMessage))
		self.assertEqual(msg.counter, 0x1)
		self.assertEqual(msg.flag, 0x00)
		self.assertEqual(msg.sender_id, 0x123456)
		self.assertEqual(msg.receiver_id, 0xE016C)
		self.assertEqual(msg.group_id, 0)
		self.assertEqual(msg.payload, "00")
		self.assertEqual(msg.decoded_payload, {
			'devicetype': "Cube",
		})

	def test_time_information_question(self):
		sample = "Z0A000A030E016C12345600"
		msg = MoritzMessage.decode_message(sample)
		self.assertTrue(isinstance(msg, TimeInformationMessage))
		self.assertEqual(msg.counter, 0x00)
		self.assertEqual(msg.flag, 0x0A)
		self.assertEqual(msg.sender_id, 0xE016C)
		self.assertEqual(msg.receiver_id, 0x123456)
		self.assertEqual(msg.group_id, 0)
		self.assertEqual(msg.payload, "")

	def test_time_information(self):
		sample = "Z0F0204031234560E016C000E0102E117"
		msg = MoritzMessage.decode_message(sample)
		self.assertTrue(isinstance(msg, TimeInformationMessage))
		self.assertEqual(msg.counter, 0x02)
		self.assertEqual(msg.flag, 0x04)
		self.assertEqual(msg.sender_id, 0x123456)
		self.assertEqual(msg.receiver_id, 0xE016C)
		self.assertEqual(msg.group_id, 0)
		self.assertEqual(msg.payload, "0E0102E117")
		self.assertEqual(msg.decoded_payload, datetime(2014, 12, 1, 2, 33, 23))

class MessageGeneralOutputTestCase(unittest.TestCase):
	def test_encoding_without_payload(self):
		expected_result = "Zs0AB900F11234560B355400"
		msg = WakeUpMessage()
		msg.counter = 0xB9
		msg.sender_id = 0x123456
		msg.receiver_id = 0x0B3554
		msg.group_id = 0
		encoded_message = msg.encode_message()
		self.assertEqual(encoded_message, expected_result)

	def test_encoding_with_payload(self):
		expected_result = "Zs0BB900401234560B3554004B"
		msg = SetTemperatureMessage()
		msg.counter = 0xB9
		msg.sender_id = 0x123456
		msg.receiver_id = 0x0B3554
		msg.group_id = 0
		payload = {
			'desired_temperature': 5.5,
			'mode': 'manual',
		}
		encoded_message = msg.encode_message(payload)
		self.assertEqual(encoded_message, expected_result)
		self.assertEqual(msg.payload, expected_result[-2:])

	def test_encoding_with_broken_payload(self):
		expected_result = "Zs0BB900401234560B3554004B"
		msg = SetTemperatureMessage()
		msg.counter = 0xB9
		msg.sender_id = 0x123456
		msg.receiver_id = 0x0B3554
		msg.group_id = 0
		payload = {
			'desiredtemperature': 5.5,
			'mode': 'manual',
		}
		with self.assertRaises(MissingPayloadParameterError):
			encoded_message = msg.encode_message(payload)


class MessageOutputSampleTestCase(unittest.TestCase):
	def test_set_temperature(self):
		msg = SetTemperatureMessage()
		msg.counter = 0xB9
		msg.sender_id = 0x123456
		msg.receiver_id = 0x0B3554
		msg.group_id = 0
		payload = {
			'desired_temperature': 5.5,
			'mode': 'manual',
		}
		self.assertEqual(msg.encode_message(payload=payload), "Zs0BB900401234560B3554004B")

	def test_set_timeinformation(self):
		msg = TimeInformationMessage()
		msg.counter = 0x02
		msg.sender_id = 0x123456
		msg.receiver_id = 0xE016C
		msg.group_id = 0x0
		payload = datetime(2014, 12, 1, 2, 33, 23)
		self.assertEqual(msg.encode_message(payload=payload), "Zs0F0204031234560E016C000E0102E117")


moritzprotocol/communication.py
# -*- coding: utf-8 -*-
"""
    moritzprotocol.communication
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~

    There are two communication classes available which should run in their own thread.
    CULComThread performs low-level serial communication, CULMessageThread performs high-level
    communication and spawns a CULComThread for its low-level needs.

    Generally just use CULMessageThread unless you have a good reason not to.

    :copyright: (c) 2014 by Markus Ullmann.
    :license: BSD, see LICENSE for more details.
"""

# environment constants

# python imports
from collections import defaultdict
from datetime import datetime
import Queue
import threading
import time

# environment imports
import logbook
from serial import Serial

# custom imports
from moritzprotocol.exceptions import MoritzError
from moritzprotocol.messages import (
    MoritzMessage, MoritzError,
    PairPingMessage, PairPongMessage,
    TimeInformationMessage,
    SetTemperatureMessage, ThermostatStateMessage, AckMessage
)
from moritzprotocol.signals import thermostatstate_received, device_pair_accepted, device_pair_request

# local constants
com_logger = logbook.Logger("CUL Serial")
message_logger = logbook.Logger("CUL Messaging")

# Hardcodings based on FHEM recommendations
CUBE_ID = 0x123456
WALLTHERMO_ID = 0x123457
SHUTTERCONTACT_ID = 0x123458


class CULComThread(threading.Thread):
    """Low-level serial communication thread base"""

    def __init__(self, send_queue, read_queue, device_path):
        super(CULComThread, self).__init__()
        self.send_queue = send_queue
        self.read_queue = read_queue
        self.device_path = device_path
        self.pending_line = []
        self.stop_requested = threading.Event()
        self.cul_version = ""
        self._has_send_budget = False

    def run(self):
        self._init_cul()
        while not self.stop_requested.isSet():
            # Process pending received messages (if any)
            read_line = self._read_result()
            if read_line is not None:
                com_logger.info("Got unhandled response from CUL: '%s'" % read_line)

            # send queued messages yet respecting send budget of 1%
            if not self.send_queue.empty() and self.has_send_budget:
                com_logger.debug("Processing queued outgoing message(s)")
                while not self.send_queue.empty():
                    com_logger.debug("Checking available budget (if not already done)")
                    if not self.has_send_budget:
                        break
                    com_logger.debug("Fetching message from queue")
                    out_msg = self.send_queue.get(True, 0.05)
                    if out_msg is None:
                        com_logger.debug("Failed fetching message due to thread lock, deferring")
                        break
                    com_logger.debug("Queueing command %s" % out_msg)
                    self.send_command(out_msg)

            # give the system 250ms to do something else, we're embedded....
            time.sleep(0.25)

    def join(self, timeout=None):
        self.stop_requested.set()
        super(CULComThread, self).join(timeout)

    def _init_cul(self):
        """Ensure CUL reports reception strength and does not do FS messages"""

        self.com_port = Serial(self.device_path)
        self._read_result()
        # get CUL FW version
        self.send_command("V")
        time.sleep(0.3)
        self.cul_version = self._read_result() or ""
        # enable reporting of message strength
        self.send_command("X21")
        time.sleep(0.3)
        # receive Moritz messages
        self.send_command("Zr")
        time.sleep(0.3)
        # disable FHT mode by setting station to 0000
        self.send_command("T01")
        time.sleep(0.3)
        self._read_result()

    @property
    def has_send_budget(self):
        """Ask CUL if we have enough budget of the 1 percent rule left"""

        if self._has_send_budget:
            return self._has_send_budget
        self.send_command("X")
        for i in range(0,10):
            result = self._read_result()
            if result is None:
                time.sleep(0.2)
                continue
            if result[0:2] != "21":
                # we set X21 in the beginning and this should be our response now
                com_logger.debug("Received unrelated message for budget question: '%s'" % result)
                continue
            remaining_ms = int(result[3:].strip()) * 10
            if remaining_ms > 2000:
                self._has_send_budget = True
                com_logger.debug("Enough send budget: %s ms" % remaining_ms)
            else:
                com_logger.info("Currently no send budget. Only %s ms available and we need at least 2000 ms")
                break
        return self._has_send_budget

    def send_command(self, command):
        """Sends given command to CUL. Invalidates has_send_budget if command starts with Zs"""

        if command.startswith("Zs"):
            self._has_send_budget = False
        self.com_port.write(command + "\r\n")
        com_logger.debug("sent: %s" % command)

    def _read_result(self):
        """Reads data from port, if it's a Moritz message, forward directly, otherwise return to caller"""

        while self.com_port.inWaiting():
            self.pending_line.append(self.com_port.read(1))
            if self.pending_line[-1] == "\n":
                # remove newlines at the end
                completed_line = "".join(self.pending_line[:-2])
                com_logger.debug("received: %s" % completed_line)
                self.pending_line = []
                if completed_line.startswith("Z"):
                    self.read_queue.put(completed_line)
                else:
                    return completed_line


class CULMessageThread(threading.Thread):
    """High level message processing"""

    def __init__(self, command_queue, device_path):
        super(CULMessageThread, self).__init__()
        self.command_queue = command_queue
        self.thermostat_states = defaultdict(dict)
        self.thermostat_states_lock = threading.Lock()
        self.com_send_queue = Queue.Queue()
        self.com_receive_queue = Queue.Queue()
        self.com_thread = CULComThread(self.com_send_queue, self.com_receive_queue, device_path)
        self.stop_requested = threading.Event()
        self.pair_as_cube = True
        self.pair_as_wallthermostat = False
        self.pair_as_ShutterContact = False

    def run(self):
        self.com_thread.start()
        while not self.stop_requested.isSet():
            message = None
            try:
                received_msg = self.com_receive_queue.get(True, 0.05)
                message = MoritzMessage.decode_message(received_msg[:-2])
                signal_strength = int(received_msg[-2:], base=16)
                self.respond_to_message(message, signal_strength)
            except Queue.Empty:
                pass
            except MoritzError as e:
                message_logger.error("Message parsing failed, ignoring message '%s'. Reason: %s" % (received_msg, str(e)))

            try:
                msg, payload = self.command_queue.get(True, 0.05)
                raw_message = msg.encode_message(payload)
                message_logger.debug("send type %s" % msg)
                self.com_send_queue.put(raw_message)
            except Queue.Empty:
                pass

            time.sleep(0.3)

    def join(self, timeout=None):
        self.com_thread.join(timeout)
        self.stop_requested.set()
        super(CULMessageThread, self).join(timeout)

    def respond_to_message(self, msg, signal_strenth):
        """Internal function to respond to incoming messages where appropriate"""

        if isinstance(msg, PairPingMessage):
            message_logger.info("received PairPing")
            # Some peer wants to pair. Let's see...
            device_pair_request.send(self, msg=msg)
            if msg.receiver_id == 0x0:
                # pairing after factory reset
                if not (self.pair_as_cube or self.pair_as_wallthermostat or self.pair_as_ShutterContact):
                    message_logger.info("Pairing to new device but we should ignore it")
                    return
                resp_msg = PairPongMessage()
                resp_msg.counter = 1
                resp_msg.sender_id = CUBE_ID
                resp_msg.receiver_id = msg.sender_id
                resp_msg.group_id = msg.group_id
                message_logger.info("responding to pair after factory reset")
                self.command_queue.put((resp_msg, {"devicetype": "Cube"}))
                device_pair_accepted.send(self, resp_msg=resp_msg)
            elif msg.receiver_id == CUBE_ID:
                # pairing after battery replacement
                resp_msg = PairPongMessage()
                resp_msg.counter = 1
                resp_msg.sender_id = CUBE_ID
                resp_msg.receiver_id = msg.sender_id
                resp_msg.group_id = msg.group_id
                message_logger.info("responding to pair after battery replacement")
                self.command_queue.put((resp_msg, {"devicetype": "Cube"}))
                device_pair_accepted.send(self, resp_msg=resp_msg)
            else:
                # pair to someone else after battery replacement, don't care
                message_logger.info("pair after battery replacement sent to other device 0x%X, ignoring" % msg.receiver_id)
                return

        elif isinstance(msg, TimeInformationMessage):
            if not msg.payload and msg.receiver_id == CUBE_ID:
                # time information requested
                resp_msg = TimeInformationMessage()
                resp_msg.counter = 1
                resp_msg.sender_id = CUBE_ID
                resp_msg.receiver_id = msg.sender_id
                resp_msg.group_id = msg.group_id
                message_logger.info("time information requested by 0x%X, responding" % msg.sender_id)
                self.command_queue.put((resp_msg, datetime.now()))

        elif isinstance(msg, ThermostatStateMessage):
            with self.thermostat_states_lock:
                message_logger.info("thermostat state updated for 0x%X" % msg.sender_id)
                self.thermostat_states[msg.sender_id].update(msg.decoded_payload)
                self.thermostat_states[msg.sender_id]['last_updated'] = datetime.now()
                self.thermostat_states[msg.sender_id]['signal_strenth'] = signal_strenth
            thermostatstate_received.send(self, msg=msg)

        elif isinstance(msg, AckMessage):
            if msg.receiver_id == CUBE_ID and msg.decoded_payload["state"] == "ok":
                thermostatstate_received.send(self, msg=msg)
                with self.thermostat_states_lock:
                    message_logger.info("ack and thermostat state updated for 0x%X" % msg.sender_id)
                    self.thermostat_states[msg.sender_id].update(msg.decoded_payload)
                    self.thermostat_states[msg.sender_id]['last_updated'] = datetime.now()
                    self.thermostat_states[msg.sender_id]['signal_strenth'] = signal_strenth


moritzprotocol/messages.py
# -*- coding: utf-8 -*-
"""
    moritzprotocol.messages
    ~~~~~~~~~~~~~~~~~~~~~~~

    Definition of known messages, based on IDs from FHEM plugin

    :copyright: (c) 2014 by Markus Ullmann.
    :license: BSD, see LICENSE for more details.
"""

# environment constants

# python imports
from datetime import datetime
import struct

# environment imports

# custom imports
from moritzprotocol.exceptions import (
	MoritzError, LengthNotMatchingError,
	MissingPayloadParameterError, UnknownMessageError
)

# local constants
DEVICE_TYPES = {
  0: "Cube",
  1: "HeatingThermostat",
  2: "HeatingThermostatPlus",
  3: "WallMountedThermostat",
  4: "ShutterContact",
  5: "PushButton"
}
DEVICE_TYPES_BY_NAME = dict((v,k) for k, v in DEVICE_TYPES.items())

MODE_IDS = {
	0: "auto",
	1: "manual",
	2: "temporary",
	3: "boost",
}


class MoritzMessage(object):
	"""Represents (de)coded message as seen on Moritz Wire"""

	def __init__(self):
		self.counter = 0
		self.flag = 0
		self.sender_id = 0
		self.receiver_id = 0
		self.group_id = 0
		self.payload = ""

	@property
	def decoded_payload(self):
		raise NotImplementedError()

	@property
	def is_broadcast(self):
		return self.receiver_id == 0

	@staticmethod
	def decode_message(input_string):
		"""Decodes given message and returns content in matching message class"""

		if input_string.startswith("Zs"):
			# outgoing messages can be parsed too, just cut the Z off as it doesn't matter
			input_string = input_string[1:]

		# Split MAX message
		length = int(input_string[1:3], base=16)
		counter = int(input_string[3:5], base=16)
		flag = int(input_string[5:7], base=16)
		msgtype = int(input_string[7:9], base=16)
		sender_id = int(input_string[9:15], base=16)
		receiver_id = int(input_string[15:21], base=16)
		group_id = int(input_string[21:23], base=16)
		payload = input_string[23:]

		# Length: strlen(input_string) / 2 as HEX encoding, +3 for Z and length count
		if (len(input_string) - 3) != length * 2:
			raise LengthNotMatchingError("Message length %i not matching indicated length %i" % ((len(input_string) - 3)/2, length))

		try:
			message_class = MORITZ_MESSAGE_IDS[msgtype]
		except KeyError:
			raise UnknownMessageError("Unknown message with id %x found" % msgtype)

		message = message_class()
		message.counter = counter
		message.flag = flag
		message.group_id = group_id
		message.sender_id = sender_id
		message.receiver_id = receiver_id
		message.payload = payload

		return message

	def encode_message(self, payload={}):
		"""Prepare message to be sent on wire"""

		msg_ids = dict((v,k) for k, v in MORITZ_MESSAGE_IDS.items())
		msg_id = msg_ids[self.__class__]

		message = ""
		if hasattr(self, 'encode_payload'):
			self.payload = self.encode_payload(payload)
		if hasattr(self, 'encode_flag'):
			self.flag = self.encode_flag()
		for (var, length) in ((self.counter,2), (self.flag,2), (msg_id, 2), (self.sender_id,6), (self.receiver_id,6), (self.group_id,2)):
			content = "%X".upper() % var
			message += content.zfill(length)
		if self.payload:
			message += self.payload
		length = "%X".upper() % (len(message)/2)
		message = "Zs" + length.zfill(2) + message
		return message

	def __repr__(self):
		return "<%s counter:%x flag:%x sender:%x receiver:%x group:%x payload:%s>" % (
			self.__class__.__name__, self.counter, self.flag, self.sender_id, self.receiver_id, self.group_id, self.payload
		)


class PairPingMessage(MoritzMessage):
	"""Thermostats send this request on long boost keypress"""

	@property
	def decoded_payload(self):
		firmware_version, device_type, selftest_result = struct.unpack(">bBB", bytearray.fromhex(self.payload[:6]))
		device_serial = self.payload[6:]
		result = {
			'firmware_version': "V%i.%i" % (firmware_version/0x10, firmware_version % 0x10),
			'device_type': DEVICE_TYPES[device_type],
			'selftest_result': selftest_result,
			'device_serial': "".join(struct.unpack(">" + "c"*(len(device_serial)/2), bytearray.fromhex(device_serial))),
			'pairmode': 'pair' if self.is_broadcast else 're-pair'
		}
		return result


class PairPongMessage(MoritzMessage):
	"""Awaited after PairPingMessage is sent by component"""

	@property
	def decoded_payload(self):
		return {'devicetype': DEVICE_TYPES[int(self.payload)]}

	def encode_payload(self, payload):
		return str(DEVICE_TYPES_BY_NAME[payload['devicetype']]).zfill(2)


class AckMessage(MoritzMessage):
	"""Last command received and acknowledged.
	   Occasionally if the communication is ongoing, this might get lost.
	   So don't rely on it but check state afterwards instead"""

	@property
	def decoded_payload(self):
		result = {}
		if self.payload.startswith("01"):
			result["state"] = "ok"
		elif self.payload.startswith("81"):
			result["state"] = "invalid_command"
		if len(self.payload) == 8:
			# FIXME: temporarily accepting the fact that we only handle Thermostat results
			result.update(ThermostatStateMessage.decode_status(self.payload[2:]))
		return result


class TimeInformationMessage(MoritzMessage):
	"""Current time is either requested or encoded. Request simply is empty payload"""

	@property
	def decoded_payload(self):
		(years_since_200, day, hour, month_minute, month_sec) = struct.unpack(">BBBBB", bytearray.fromhex(self.payload[:12]))
		return datetime(
			year=years_since_200 + 2000,
			minute=month_minute & 0x3F,
			month=((month_minute >> 4) & 0x0C) | ((month_sec >> 6) & 0x03),
			day=day,
			hour=hour,
			second=month_sec & 0x3F
		)

	def encode_flag(self):
		return 0x0A if not self.payload else 0x04

	def encode_payload(self, payload=None):
		# may contain empty payload to ask for timeinformation
		if payload is None:
			return ""
		encoded_payload = str("%X" % (payload.year - 2000)).zfill(2)
		encoded_payload += str("%X" % payload.day).zfill(2)
		encoded_payload += str("%X" % payload.hour).zfill(2)
		encoded_payload += str("%X" % (payload.minute | ((payload.month & 0x0C) << 4))).zfill(2)
		encoded_payload += str("%X" % (payload.second | ((payload.month & 0x03) << 6))).zfill(2)
		return encoded_payload


class ConfigWeekProfileMessage(MoritzMessage):
	pass


class ConfigTemperaturesMessage(MoritzMessage):
	pass


class ConfigValveMessage(MoritzMessage):
	pass


class AddLinkPartnerMessage(MoritzMessage):
	pass


class RemoveLinkPartnerMessage(MoritzMessage):
	pass


class SetGroupIdMessage(MoritzMessage):
	pass


class RemoveGroupIdMessage(MoritzMessage):
	pass


class ShutterContactStateMessage(MoritzMessage):
	pass


class SetTemperatureMessage(MoritzMessage):
	"""Sets temperature for manual mode as well as mode switch between manual, auto and boost"""

	@property
	def decoded_payload(self):
		payload = struct.unpack(">B", bytearray.fromhex(self.payload[0:4]))
		return {
			'desired_temperature': ((payload[0] & 0x3F) / 2.0),
			'mode': MODE_IDS[payload[0] >> 6]
		}

	def encode_flag(self):
		return 0x4 if self.group_id else 0x0

	def encode_payload(self, payload):
		if "desired_temperature" not in payload:
			raise MissingPayloadParameterError("Missing desired_temperature in payload")
		if "mode" not in payload:
			raise MissingPayloadParameterError("Missing mode in payload")

		if payload['desired_temperature'] > 30.5:
			desired_temperature = 30.5 # "ON"
		elif payload['desired_temperature'] < 4.5:
			desired_temperature = 4.5 # "OFF"
		else:
			#always round to nearest 0.5 first
			desired_temperature = round(payload['desired_temperature']*2)/2.0
		int_temperature = int(desired_temperature*2)

		modes = dict((v,k) for (k,v) in MODE_IDS.items())
		mode = modes[payload['mode']]

		content = "%X".upper() % ((mode << 6) | int_temperature)
		return content.zfill(2)


class WallThermostatControlMessage(MoritzMessage):
	pass


class SetComfortTemperatureMessage(MoritzMessage):
	pass


class SetEcoTemperatureMessage(MoritzMessage):
	pass


class PushButtonStateMessage(MoritzMessage):
	pass


class ThermostatStateMessage(MoritzMessage):
	"""Non-reculary sent by Thermostats to report when valve was moved or command received."""

	@staticmethod
	def decode_status(payload):
		status_bits, valve_position, desired_temperature = struct.unpack(">bBB", bytearray.fromhex(payload[0:6]))
		mode = status_bits & 0x3
		dstsetting = status_bits & 0x04
		langateway = status_bits & 0x08
		status_bits = status_bits >> 9
		is_locked = status_bits & 0x1
		rferror = status_bits & 0x2
		battery_low = status_bits & 0x4
		desired_temperature = (desired_temperature & 0x7F) / 2.0
		result = {
			"mode": MODE_IDS[mode],
			"dstsetting": bool(dstsetting),
			"langateway": bool(langateway),
			"is_locked": bool(is_locked),
			"rferror": bool(rferror),
			"battery_low": bool(battery_low),
			"desired_temperature": desired_temperature,
			"valve_position": valve_position,
		}
		return result

	@property
	def decoded_payload(self):
		result = ThermostatStateMessage.decode_status(self.payload)
		if len(self.payload) > 6:
			pending_payload = bytearray.fromhex(self.payload[6:])
			if len(pending_payload) == 3:
				# TODO handle date string
				pass
			elif len(pending_payload) == 2 and result['mode'] != 'temporary':
				result["measured_temperature"] = (((pending_payload[0] & 0x1) << 8) + pending_payload[1]) / 10.0
			else:
				# unknown....
				pass
		return result


class WallThermostatStateMessage(MoritzMessage):
	pass


class SetDisplayActualTemperatureMessage(MoritzMessage):
	pass


class WakeUpMessage(MoritzMessage):
	pass


class ResetMessage(MoritzMessage):
	"""Perform a factory reset on given device"""

	pass


# Define at bottom so we can use the class types right away
# Based on FHEM CUL_MAX module
MORITZ_MESSAGE_IDS = {
	0x00: PairPingMessage,
	0x01: PairPongMessage,
	0x02: AckMessage,
	0x03: TimeInformationMessage,

	0x10: ConfigWeekProfileMessage,
	0x11: ConfigTemperaturesMessage,
	0x12: ConfigValveMessage,

	0x20: AddLinkPartnerMessage,
	0x21: RemoveLinkPartnerMessage,
	0x22: SetGroupIdMessage,
	0x23: RemoveGroupIdMessage,

	0x30: ShutterContactStateMessage,

	0x40: SetTemperatureMessage,
	0x42: WallThermostatControlMessage,
	0x43: SetComfortTemperatureMessage,
	0x44: SetEcoTemperatureMessage,

	0x50: PushButtonStateMessage,

	0x60: ThermostatStateMessage,

	0x70: WallThermostatStateMessage,

	0x82: SetDisplayActualTemperatureMessage,

	0xF1: WakeUpMessage,
	0xF0: ResetMessage,
}
